lib/asciidoctor/js/opal_ext.rb

```diff
@@ -1,121 +1,8 @@
-      engine = engine || 'v8';
-      if (isElectron) {
-        framework = framework || 'electron';
-      }
-    }
-    else if (isNashorn) {
-      platform = platform || 'java';
-      engine = engine || 'nashorn';
-    }
-    else if (isRhino) {
-      platform = platform || 'java';
-      engine = engine || 'rhino';
-    }
-    else if (isSpiderMonkey) {
-      platform = platform || 'standalone';
-      framework = framework || 'spidermonkey';
-    }
-    else if (isBrowser) {
-      platform = platform || 'browser';
-      if (isPhantomJS) {
-        framework = framework || 'phantomjs';
-      }
-    }
-    // NOTE: WebWorker are not limited to browser
-    if (isWebWorker) {
-      framework = framework || 'webworker';
-    }
-  }
-
-  if (typeof platform === 'undefined') {
-    throw new Error('Unable to automatically detect the JavaScript platform, please configure Asciidoctor.js: `Asciidoctor({runtime: {platform: \'node\'}})`');
-  }
-
-  // Optional information
-  if (typeof framework === 'undefined') {
-    framework = '';
-  }
-  if (typeof engine === 'undefined') {
-    engine = '';
-  }
-
-  // IO Module
-  if (typeof ioModule !== 'undefined') {
-    if (ioModule != 'spidermonkey'
-         && ioModule != 'phantomjs'
-         && ioModule != 'node'
-         && ioModule != 'java_nio'
-         && ioModule != 'xmlhttprequest') {
-      throw new Error('Invalid IO module, `config.ioModule` must be one of: spidermonkey, phantomjs, node, java_nio or xmlhttprequest');
-    }
-  } else {
-    if (framework === 'spidermonkey') {
-      ioModule = 'spidermonkey';
-    } else if (framework === 'phantomjs') {
-      ioModule = 'phantomjs';
-    } else if (platform === 'node') {
-      ioModule = 'node';
-    } else if (engine === 'nashorn') {
-      ioModule = 'java_nio'
-    } else if (platform === 'browser' || typeof XmlHTTPRequest !== 'undefined') {
-      ioModule = 'xmlhttprequest'
-    } else {
-      throw new Error('Unable to automatically detect the IO module, please configure Asciidoctor.js: `Asciidoctor({runtime: {ioModule: \'node\'}})`');
-    }
-  }
-
-)
-
-JAVASCRIPT_IO_MODULE = %x(ioModule)
-JAVASCRIPT_PLATFORM = %x(platform)
-JAVASCRIPT_ENGINE = %x(engine)
-JAVASCRIPT_FRAMEWORK = %x(framework)
-
 require 'asciidoctor/js/opal_ext/file'
 require 'asciidoctor/js/opal_ext/match_data'
 require 'asciidoctor/js/opal_ext/kernel'
 require 'asciidoctor/js/opal_ext/thread_safe'
 require 'asciidoctor/js/opal_ext/string'
 require 'asciidoctor/js/opal_ext/uri'
 
-if JAVASCRIPT_ENGINE == 'nashorn'
-  require 'asciidoctor/js/opal_ext/nashorn/io'
-  require 'asciidoctor/js/opal_ext/nashorn/dir'
-end
-if JAVASCRIPT_FRAMEWORK == 'electron'
-  require 'asciidoctor/js/opal_ext/electron/io'
-end
-if JAVASCRIPT_PLATFORM == 'node'
-  require 'asciidoctor/js/opal_ext/node/file'
-  `Opal.load("nodejs")`
-end
+require 'asciidoctor/js/opal_ext/env' # dirty hack, the require will be replaced by a specific implementation
```

Comment:
The goal is to transpile Asciidoctor only once and then to transpile each specific implementation.
So we end up with: 

* `asciidoctor-core.js` which contains `asciidoctor` and `asciidoctor/js`
* `opal-ext-browser.js` which contains `asciidoctor/js/opal_ext/browser`
* `opal-ext-node.js` which contains `asciidoctor/js/opal_ext/node`
* `opal-ext-umd.js` which contains `asciidoctor/js/opal_ext/umd`
* `opal-ext-nashorn.js` which contains `asciidoctor/js/opal_ext/nashorn`
* ...

The tricky part is to require the specific implementation (at the right time).
So my idea was to replace a "fake" require and concat the two files. For instance if we want to generate Asciidoctor.js for the browser environment:

1. Replace `require 'asciidoctor/js/opal_ext/env'` by `require 'asciidoctor/js/opal_ext/browser`
1. Concat `opal-ext-browser.js` with `asciidoctor-core.js`
1. Enjoy 😋 

But I'm not satisfied with this solution... @mojavelinux Do you have a better idea ?

Reply:
I agree we don't want this to be a runtime require. What we want is something very similar to browserify or webpack. Instead of a dedicated name, though, it might be better to use some sort of symbol. I've seen this done in Node using something like a `~` prefix or a magic name like `__asciidoctor_env__`.

Reply:
I'm now using a "magic" code comment instead of a require on an empty file.

> I agree we don't want this to be a runtime require. 

Actually we need a runtime require because Opal is generating a module for each runtime, so we have to require the module. We could also use an auto requirable module... but for now I think it's OK.

Reply:
:+1: